packages/@lwc/template-compiler/src/__tests__/fixtures/directive-lwc-if-else/dynamic/if-elseif-multiple/expected.js

```diff
@@ -0,0 +1,29 @@
+import { parseFragment, registerTemplate } from "lwc";
+const $fragment1 = parseFragment`<h1${3}>if condition</h1>`;
+const stc0 = {
+  key: 2,
+};
+const stc1 = {
+  key: 3,
+};
+const stc2 = {
+  key: 4,
+};
+function tmpl($api, $cmp, $slotset, $ctx) {
+  const {
+    st: api_static_fragment,
+    t: api_text,
+    h: api_element,
+    dc: api_dynamic_component,
+  } = $api;
+  return $cmp.visible.if
+    ? [api_static_fragment($fragment1(), 1)]
```

Comment:
@jodarove You're correct, that code does throw. I've added a simple karma test in the latest commit.

I'm really unfamiliar with this area, but does this mean these children need to be treated as dynamic children when going through the diffing algo?

cc - @jmsjtu

Reply:
Alright, so after digging into this a bit, it looks like the engine doesn't have a mechanism to just "replace" the old nodes with the new nodes like previously thought. For the existing `if:true` and `if:false`, two separate VNodes are generated independently, and the patching is treated as a remove + add per [the logic here](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L762).

I think we need to spend some more time thinking what our diffing assumptions/requirements are for if/elseif/else branches.

If we're okay with no diffing, similar to the existing `if:true` `if:false` logic, we could wrap the entire ternary operation in a new type of VNode so that the engine knows it's okay to unmount all the old children and mount the new children.  Something like the following:
```typescript
patchIfFragment(n1: IfVFragment, n2: IfVFragment, parent: ParentNode, renderer: RendererAPI) {
    // ...?
    unmountVNodes(n1.children, parent, renderer, null);
    mountVNodes(n2.children, parent, renderer, null);
}
```
The `VFragment` work in https://github.com/salesforce/lwc/pull/2957 could help with this, but from what I can see, that VFragment still leads us through the same diffing logic that eventually tries to patch two different vnodes.

Does this sound right?

Reply:
> I'm really unfamiliar with this area, but does this mean these children need to be treated as dynamic children when going through the diffing algo?

yes, we need these children to be treated as dynamic children when going through the diffing algo. By doing so, the diffing algo won't try to patch 2 different nodes, instead will remove the old and insert the new.

> the engine doesn't have a mechanism to just "replace" the old nodes with the new nodes like previously thought

The mechanism to "replace" is like you described to delete the old, and insert the new. that is only achieved in the `updateDynamicChildren` routine.  notice that is done when patching all the children (is not done individually)

The `updateStaticChildren`, assumes that between rendering cycles, at the same position of children, it will be null (it will remove the old or insert the new) or the same (updated) vnode, in this case it will patch the old with the new. but never to "replace".

> If we're okay with no diffing, similar to the existing if:true if:false logic, we could wrap the entire ternary operation in a new type of VNode so that the engine knows it's okay to unmount all the old children and mount the new children.

The way i see it, the unmount+mount you describe, is the same as going through the `updateDynamicChildren`. The only benefit in that case, is that it won't punish the siblings of the if/elseif/else, by forcing them to also go through the slower `updateDynChildren`. That is one of the benefits of the `VFragment` work

> The VFragment work in https://github.com/salesforce/lwc/pull/2957 could help with this, but from what I can see, that VFragment still leads us through the same diffing logic that eventually tries to patch two different vnodes.
Does this sound right?

It will go through the same logic, but it won't try to patch two different vnodes, because `updateDynamicChildren` checks before patching that the two vnodes are the same.